src/ray/common/memory_monitor.cc

```diff
@@ -0,0 +1,146 @@
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/common/memory_monitor.h"
+
+#include <fstream>  // std::ifstream
+#include <tuple>
+
+#include "ray/common/ray_config.h"
+#include "ray/util/logging.h"
+
+namespace ray {
+
+MemoryMonitor::MemoryMonitor(float usage_threshold_,
+                             uint64_t monitor_interval_ms,
+                             MemoryUsageRefreshCallback monitor_callback)
+    : usage_threshold_(usage_threshold_),
+      monitor_callback_(monitor_callback),
+      io_context_(),
+      monitor_thread_([this] {
+        boost::asio::io_service::work io_service_work_(io_context_);
+        io_context_.run();
+      }),
+      runner_(io_context_) {
+  RAY_CHECK(monitor_callback_ != nullptr);
+  RAY_CHECK(usage_threshold_ >= 0);
+  RAY_CHECK(usage_threshold_ <= 1);
+  if (monitor_interval_ms > 0) {
+#ifdef __linux__
+    runner_.RunFnPeriodically(
+        [this] {
+          bool is_usage_above_threshold = IsUsageAboveThreshold();
+          monitor_callback_(is_usage_above_threshold);
+        },
+        monitor_interval_ms,
+        "MemoryMonitor.CheckIsMemoryUsageAboveThreshold");
+    RAY_LOG(INFO) << "MemoryMonitor initialized";
+#else
+    RAY_LOG(WARNING) << "Not running MemoryMonitor. It is currently supported "
+                     << "only on Linux.";
+#endif
+  } else {
+    RAY_LOG(INFO) << "MemoryMonitor disabled. Specify "
+                  << "`memory_monitor_interval_ms` > 0 to enable the monitor.";
+  }
+}
+
+bool MemoryMonitor::IsUsageAboveThreshold() {
+  auto [used_memory_bytes, total_memory_bytes] = GetLinuxNodeMemoryBytes();
+  if (total_memory_bytes == 0) {
+    RAY_LOG(ERROR) << "Unable to capture node memory. Monitor will not be able "
+                   << "to detect memory usage above threshold.";
+    return false;
+  }
+  auto usage_fraction = static_cast<float>(used_memory_bytes) / total_memory_bytes;
+  bool is_usage_above_threshold = usage_fraction > usage_threshold_;
+  if (is_usage_above_threshold) {
+    RAY_LOG_EVERY_MS(INFO, 1000)
+        << "Node memory usage above threshold, used: " << used_memory_bytes
+        << ", total: " << total_memory_bytes << ", usage fraction: " << usage_fraction
+        << ", threshold: " << usage_threshold_;
+  }
+  return is_usage_above_threshold;
+}
+
+std::tuple<uint64_t, uint64_t> MemoryMonitor::GetLinuxNodeMemoryBytes() {
+#ifndef __linux__
+  RAY_LOG(ERROR) << "GetLinuxNodeMemoryBytes called on non-Linux system" return return {
+      0, 0};
+#endif
+  std::string meminfo_path = "/proc/meminfo";
```

Comment:
Can we use `sysinfo` https://man7.org/linux/man-pages/man2/sysinfo.2.html ?

Reply:
Tried sysinfo and noticed it has buffer, free, shared, and total only.
For us to calculate used we need either the used or available column, that is only available in free / meminfo, etc.